### ISSUE TITLE
Replace expired aad credentials in cluster

### ISSUE THREAD

**Issue by gaieges** [Azure/aks-engine#4596]
**Describe the bug**

Not able to provision any azure resources (eg PV) from a given cluster, due to the following error:

```
Warning  ProvisioningFailed  10m                  persistentvolume-controller  Failed to provision volume with StorageClass "default": azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://mana │
│ gement.azure.com/subscriptions/REDACTED/resourceGroups/REDACTED/providers/Microsoft.Compute/disks/REDACTED?api-version= │
│ 2018-09-30: StatusCode=401 -- Original Error: adal: Refresh request failed. Status Code = '401'. Response body: {"error":"invalid_client","error_description":"REDACTED: The provided client secret keys are expired. Visit the Azure │
│  Portal to create new keys for your app, or consider using certificate credentials for added security: https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials\r\nTrace ID: REDACTED1\r\nCorrelation ID:\r\nTimestamp: 2021-07-21 15:04:12Z","error_codes":[7000222],"timestamp":"2021-07-21 15:04:12Z","trace_id":"REDACTED1","correlation_id │
│ ":"REDACTED","error_uri":"https://login.microsoftonline.com/error?code=7000222"}      
```

Similar to issue here: https://github.com/Azure/AKS/issues/222

Is the right way to replace this aad Credential on the aks-engine cluster as described above or is there a different way?

cc: @facu 

**Steps To Reproduce**
Have an aks-engine cluster for >1 year with a service principal that expires before that.

**Expected behavior**
Cluster can provision volumes

**AKS Engine version**
v0.53.0

**Kubernetes version**
v1.16.11

**Additional context**

**Comment by rodriguez-facundo** [comment]
looks expired 

```
➜ az ad sp credential list --id "c24da16c-0083-434a-9502-5cbccb219e85" --query "[].endDate" -o tsv
2021-05-29T17:32:11.451916+00:00
```

**Comment by rodriguez-facundo** [comment]
I can't see anything running

```
az aks show --resource-group aks-engine-eastus-primary-14 --name aks-engine-eastus-primary-14-k8s
```

do I need extra permissions or aks is empty?

**Comment by rodriguez-facundo** [comment]
Ok, the problem seems to be how to update the vmss with new Service Principal credentials. I'm tempted to renew the credentials, and then recycled the vms to see if the new secret is picked up.

each vm stores it under /etc/kubernetes/azure.json

**Comment by jadarsie** [comment]
yes, update `/etc/kubernetes/azure.json` and make sure to force a restart of the containers

**Comment by gaieges** [comment]
How will new nodes that get scaled in get that update @jadarsie ?

**Comment by jadarsie** [comment]
you using cluster-autoscaler to scale up/down?

**Comment by gaieges** [comment]
No just a VMSS nodepool that has a autoscaler set up on it in azure

**Comment by jadarsie** [comment]
Never tried this myself, proceed with caution.

Executing an `aks-engine scale` command (after you update `apimodel.json` with the new creds) should update the azure deployment. After that, I imagine the VMSS will use the updated deployment to create new nodes.

**Comment by rodriguez-facundo** [comment]
Hi 👋, had no luck with the approach described ^^. Updated sp secret (gave us 1 more year), then updated apimodel.json, then `aks-engine scale` and the new VMs were created with the right values under /etc/kubernetes/azure.json but pods spawned in the new VMs keep complaining about the same issue.

**Comment by devigned** [comment]
Perhaps, that will provide time to migrate to Cluster API, which provides a much improved day 2 ops experience.

**Comment by gaieges** [comment]
@devigned can you elaborate? What can we do with this current cluster that has this issue still?

**Comment by devigned** [comment]
> @devigned can you elaborate? What can we do with this current cluster that has this issue still?

Cluster API will not help much with the current cluster, as there is no way to "import" an existing cluster to be managed by Cluster API. However, if you were to build future clusters with Cluster API (https://github.com/kubernetes-sigs/cluster-api-provider-azure), you would be able to [change the AzureClusterIdentity reference](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/a635f6a47853789a6f9c63ed256873caa8727249/templates/cluster-template.yaml#L210-L213), or change [machine deployments or pools independently](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/a635f6a47853789a6f9c63ed256873caa8727249/templates/cluster-template-user-assigned-identity.yaml#L134-L135) through K8s custom resources.

**Comment by gaieges** [comment]
Ah gotcha, understood.  Still need to do something with the current cluster, if y'all have any ideas on how to update that SP

**Comment by jadarsie** [comment]
> Hi 👋, had no luck with the approach described ^^. Updated sp secret (gave us 1 more year), then updated apimodel.json, then `aks-engine scale` and the new VMs were created with the right values under /etc/kubernetes/azure.json but pods spawned in the new VMs keep complaining about the same issue.

if `/etc/kubernetes/azure.json` has the expected credentials, maybe the pods are fetching them from a different source.

**Comment by jadarsie** [comment]
`aks-engine upgrade` will recreate the VMs (you can "upgrade" to the current k8s version), that's something else you can try, not sure if VMSS autoscaler will interfere though

**Comment by gaieges** [comment]
> > Hi 👋, had no luck with the approach described ^^. Updated sp secret (gave us 1 more year), then updated apimodel.json, then `aks-engine scale` and the new VMs were created with the right values under /etc/kubernetes/azure.json but pods spawned in the new VMs keep complaining about the same issue.
> 
> if `/etc/kubernetes/azure.json` has the expected credentials, maybe the pods are fetching them from a different source.

Do you know where they'd be fetching them from at this point given the error above?  Where does it end up getting set / used?

**Comment by vladimirjk** [comment]
Hello! Just to follow-up on the above. In order to fix the issue with SP updates just scaling down/up nodes is not enough. You have to update master nodes as well. The most suitable way to do that I think is to perform `aks-engine upgrade --force`.
